README.md
# SystemVerilog Browser package

> Still in Alpha phase! Not stable can crash when putting on panel with Project.

This package, helps to track long systemverilog files. Current Features:
* Finds all class definitions in current file
* Finds all tasks/functions
* Supports nested SystemVerilog elements
* Jumping and highlighting of the class,task and function definition by clicking on the name of the element in SystemVerilog browser window
* Clicking on the heading will clear any highlighting made with SystemVerilog Browser
* Hotkey for toggle on/off are `ctrl` + `alt` + `s`
* Added some snippets for UVM
* Experimental Implementation of Project Module Based tree hierarchy(for classes and packages).

There is a lot of bugs, no test implemented yet.

## Current File Browsing(Default)


## For Project browsing
* You need to have all the files in one root folder, the directory search is not yet implemented
* The search firstly finds *modules* then imported *packages* and then *included* files in the *packages*
* The current file browsing doesn't link other files.

TO-DO:
* Search for non wrapping SV elements (logic,bit,mailbox etc.)
* Search for occurences in SV code
* Save the state of the Panel, on save it is created again thus opening or collapsed elements
* Search for SV code in project directories, now just in root directory
* Implement object SystemVerilogTree for searching in SystemVerilog project
* More effective searching through folders
* Add packages element to the end of modules
* Adds support for extern functions/tasks


CHANGELOG.md
## 1.0.1 - First fix of the First Release
* Fixed include"somefile.sv", that wasn't added to the browser.
* Added project search for subfolders of the project.

## 1.0.0 - First Release
* Publishing on Atom repository, no changes made from the last release.

## 0.2.0 - Third Alpha Release
* Added module oriented hierarchy of project folder modules -> import statements -> package -> classes and other contents
* Added buttons to switch between curent file browsing(default) and project browsing

## 0.1.0 - Second Alpha Release
* Graphical Overhaul
* Reworked search engine, based on hierarchical cascading of objects
* Filtering commented lines

## 0.0.2 - Fix Version
* Some Text completed for README, LICENCE

## 0.0.1 - First Alpha Release
* Basic Functionality added


spec/systemverilog-browser-spec.js
'use babel';

import SystemverilogBrowser from '../lib/systemverilog-browser';

// Use the command `window:run-package-specs` (cmd-alt-ctrl-p) to run specs.
//
// To run a specific `it` or `describe` block add an `f` to the front (e.g. `fit`
// or `fdescribe`). Remove the `f` to unfocus the block.

describe('SystemverilogBrowser', () => {
  let workspaceElement, activationPromise;

  beforeEach(() => {
    workspaceElement = atom.views.getView(atom.workspace);
    activationPromise = atom.packages.activatePackage('systemverilog-browser');
  });

  describe('when the systemverilog-browser:toggle event is triggered', () => {
    it('hides and shows the modal panel', () => {
      // Before the activation event the view is not on the DOM, and no panel
      // has been created
      expect(workspaceElement.querySelector('.systemverilog-browser')).not.toExist();

      // This is an activation event, triggering it will cause the package to be
      // activated.
      atom.commands.dispatch(workspaceElement, 'systemverilog-browser:toggle');

      waitsForPromise(() => {
        return activationPromise;
      });

      runs(() => {
        expect(workspaceElement.querySelector('.systemverilog-browser')).toExist();

        let systemverilogBrowserElement = workspaceElement.querySelector('.systemverilog-browser');
        expect(systemverilogBrowserElement).toExist();

        let systemverilogBrowserPanel = atom.workspace.panelForItem(systemverilogBrowserElement);
        expect(systemverilogBrowserPanel.isVisible()).toBe(true);
        atom.commands.dispatch(workspaceElement, 'systemverilog-browser:toggle');
        expect(systemverilogBrowserPanel.isVisible()).toBe(false);
      });
    });

    it('hides and shows the view', () => {
      // This test shows you an integration test testing at the view level.

      // Attaching the workspaceElement to the DOM is required to allow the
      // `toBeVisible()` matchers to work. Anything testing visibility or focus
      // requires that the workspaceElement is on the DOM. Tests that attach the
      // workspaceElement to the DOM are generally slower than those off DOM.
      jasmine.attachToDOM(workspaceElement);

      expect(workspaceElement.querySelector('.systemverilog-browser')).not.toExist();

      // This is an activation event, triggering it causes the package to be
      // activated.
      atom.commands.dispatch(workspaceElement, 'systemverilog-browser:toggle');

      waitsForPromise(() => {
        return activationPromise;
      });

      runs(() => {
        // Now we can test for view visibility
        let systemverilogBrowserElement = workspaceElement.querySelector('.systemverilog-browser');
        expect(systemverilogBrowserElement).toBeVisible();
        atom.commands.dispatch(workspaceElement, 'systemverilog-browser:toggle');
        expect(systemverilogBrowserElement).not.toBeVisible();
      });
    });
  });
});


spec/systemverilog-browser-view-spec.js
'use babel';

import SystemverilogBrowserView from '../lib/systemverilog-browser-view';

describe('SystemverilogBrowserView', () => {
  it('has one valid test', () => {
    expect('life').toBe('easy');
  });
});


lib/system-verilog-search.js
'use babel';

import { Range, Directory, TextBuffer } from 'atom';
import SystemVerilogCodeElem from './systemverilog-file-elements';
import SystemVerilogType from './systemverilog-type';
import SystemVerilogProjectTree from './systemverilog-project-tree';

export default class SystemVerilogSearch {

    constructor(){}

    directorySearchSearching(currentDirectory,self){
      var foundFiles = [];

      try {
        var currentItems = currentDirectory.getEntriesSync();
      } catch (e) {
        console.log(e);
        return foundFiles;
      }

      currentItems.forEach(function(currentItem){
        if(currentItem.isFile() == true){
          currentFileStr = currentItem.path.match(/(?!\/)\w*(\.\w*|\w*)$/g)[0].split(".");
          if(currentFileStr[1]=="sv"||currentFileStr[1]=="svh"||currentFileStr[1]=="v"){
            foundFiles.push(self.searchBufferInPath(currentItem.path));
          }
        } else {
          retFiles = self.directorySearchSearching(currentItem,self);
          retFiles.forEach(function(retFile){foundFiles.push(retFile);});
        }
      });
      return foundFiles;
    }

    linkIncludes(tree,files){
      if(Array.isArray(tree)==true){
        tree.forEach(function(treeElement){

        });
      } else {

      }
    }

    searchProject(){
      this.project = atom.project;
      self = this;
      this.directories = [];
      paths = this.project.getPaths();
      for (var i = 0; i < Object.keys(this.project.getPaths()).length; i++) {
        this.directories.push(new Directory(paths[i]));
      }
      this.folderTree = [];
      this.directories.forEach(function (dir){
        dir.getEntriesSync().forEach(function (pathItem){
          if (pathItem.isFile() == true) {
            currentFileStr = pathItem.path.match(/(?!\/)\w*(\.\w*|\w*)$/g)[0].split(".");
            if(currentFileStr[1]=="sv"||currentFileStr[1]=="svh"||currentFileStr[1]=="v"){
              self.folderTree.push(self.searchBufferInPath(pathItem.path));
            }
          }
        });
      });
      testFiles = [];
      this.directories.forEach(function(directory){
        retFiles = self.directorySearchSearching(directory,self);
        retFiles.forEach(function(retFile){
          testFiles.push(retFile);
        });
      });
      packageTree = this.findElementsFromFolderTree("type","package",testFiles);
      // moduleStack = this.findElementsFromFolderTree("type","module",testFiles);
      // packageTree = this.findElementsFromFolderTree("type","package",this.folderTree);
      moduleStack = this.findElementsFromFolderTree("type","module",this.folderTree);

      testFiles.forEach(function(item){
        if(Object.keys(item).length !== 0){
          packageTree.forEach(function(packageItem){
            packageItem.elements.forEach(function(element){
              if(item[0].fileName==element.hardlink){
                item.forEach(function(subItem){
                  element.elements.push(subItem);
                });
              }
            });
          });
        }
      });

      modulesWithImports = [];
      moduleStack.forEach(function(item){
        var modImports = self.findElementsFromFolderTree("type","import",item);
        packageTree.forEach(function(packageFile){
          if(Array.isArray(packageFile)){
            packageFile.forEach(function(package){
              modImports.forEach(function(imports){
                retVal = self.findElementsFromFolderTree("name",imports.symlink,packageFile);
                retVal.forEach(function(val){
                  modulesWithImports.push(val);
                });
              });
            });
          }else{
            modImports.forEach(function(imports){
              retVal = self.findElementsFromFolderTree("name",imports.symlink,packageFile);
              retVal.forEach(function(val){
                val.elements.forEach(function(valElements){
                  imports.elements.push(valElements);
                });
              });
            });
          }
        });
        item.elements.forEach(function(itemElement){
          modImports.forEach(function(modImport){
            if(itemElement.fingerprint == modImport.fingerprint){
              itemElement.elements = modImport.elements;
            }
          });
        });
      });


      moduleStack.forEach(function(element){element.fixParentsInTree(null);});
      packageTree.forEach(function(element){element.fixParentsInTree(null);});

      if(Object.keys(moduleStack).length == 0){
        return packageTree;
      } else {
        return moduleStack;
      }

    }

    findElementsFromFolderTree(searchType,searchingFor,from){
      retStack = [];
      if(Array.isArray(from)){
        from.forEach(function(element){
          if(Array.isArray(element)){
            element.forEach(function(subElement){
              if (searchType == "type"){
                results = subElement.searchForType(searchingFor,[]);
              } else {
                results = subElement.searchForName(searchingFor,[]);
              }
              if(Object.keys(results).length >0){
                results.forEach(function(result){
                  retStack.push(result);
                });
              }
            });
          }else{
            if (searchType == "type"){
              results = element.searchForType(searchingFor,[]);
            } else {
              results = element.searchForName(searchingFor,[]);
            }
            if(Object.keys(results).length >0){
              results.forEach(function(result){
                retStack.push(result);
              });
            }
          }
        });
      } else {
        if (searchType == "type"){
          results = from.searchForType(searchingFor,[])
        } else {
          results = from.searchForName(searchingFor,[])
        }
        if(Object.keys(results).length >0){
          if(Array.isArray(results)){
            results.forEach(function(result){
              retStack.push(result);
            });
          }else{
            retStack.push(results);
          }
        }
      }
      return retStack;
    }

    searchCurrentEditor(){
      this.editor = atom.workspace.getActiveTextEditor();
      this.searchBufferInPath(this.editor.getPath());
      return this.localElements;
    }


    searchBufferInPath(path){
      buffer =  TextBuffer.loadSync(path);
      this.commentRangesStack = this.searchForComments(buffer);
      this.localElements = this.searchForFileElements(buffer);
      return this.localElements;
    }

    searchElement(foundElementsStack,position,svFileInfo){
      var currentParent = new SystemVerilogCodeElem(foundElementsStack[position][0],foundElementsStack[position][0],foundElementsStack[position][1],svFileInfo);
      position += 1;
      while(Object.keys(foundElementsStack).length > position){
        nextType = new SystemVerilogType(foundElementsStack[position][1]);
        if (nextType.border == 0){
          retVal = this.searchElement(foundElementsStack,position,svFileInfo);
          retVal.currentElement.parent = currentParent;
          currentParent.elements.push(retVal.currentElement);
          position = retVal.position;
        } else {
            if (nextType.withRange == 0 || nextType.extern == 1) {
              var currentChild = new SystemVerilogCodeElem(foundElementsStack[position][0],foundElementsStack[position][0],foundElementsStack[position][1],svFileInfo);
              currentChild.parent = currentParent;
              currentParent.elements.push(currentChild);
              position += 1;
            } else {
              currentParent.setEndRange(foundElementsStack[position][0]);
              position += 1;
              return {
                  currentElement: currentParent,
                  position: position
              }
          }
        }
      }
      return {
          currentElement: currentParent,
          position: position
      };
    }

    searchForFileElements(file){
      var foundElementsStack = [];
      file.scanInRange(/\b((class\s){1}(\w*)(\sextends.*;|\s;|;)|(endclass.*)$)/g, file.getRange(), event => foundElementsStack.push([event.range,event.matchText]))
      file.scanInRange(/\b((virtual\s)*(extern\s)*(virtual\s)*(function\s){1}(.*;|\s;|;)|(endfunction.*)$)/g, file.getRange(), event => foundElementsStack.push([event.range,event.matchText]))
      file.scanInRange(/\b((task\s){1}(.*;|\s;|;)|(endtask.*)$)/g, file.getRange(), event => foundElementsStack.push([event.range,event.matchText]))
      file.scanInRange(/\b((module\s.*(?!;))|(endmodule.*))$/g, file.getRange(), event => foundElementsStack.push([event.range,event.matchText]))
      file.scanInRange(/\b((interface)\s(.*;|\s;|;)|(endinterface.*)$)/g, file.getRange(), event => foundElementsStack.push([event.range,event.matchText]))
      file.scanInRange(/\b((covergroup)\s(.*;|\s;|;)|(endgroup.*)$)/g, file.getRange(), event => foundElementsStack.push([event.range,event.matchText]))
      file.scanInRange(/\b((package)\s(.*(?=;))|(endpackage.*)$)/g, file.getRange(), event => foundElementsStack.push([event.range,event.matchText]))
      file.scanInRange(/(?!\b`)include.*$/g, file.getRange(), event => foundElementsStack.push([event.range,event.matchText]))
      file.scanInRange(/(?!\b`)import.*(?=;$)/g, file.getRange(), event => foundElementsStack.push([event.range,event.matchText]))
      foundElementsStack = foundElementsStack.sort((a,b) => a[0].start.row-b[0].start.row);
      this.commentRangesStack.forEach(function(comment) {
        foundElementsStack = foundElementsStack.filter(elem => !comment.containsRange(elem[0]));
      });
      foundElementsStack.unshift([new Range([[0,0],[0,0]]),"svFile "+file.getPath()]);
      retVal = this.searchElement(foundElementsStack,0,file.getPath());
      parsedElements = retVal.currentElement.elements; // To get rid off svFile
      parsedElements.forEach(function(element){
        element.parent = null;
      });
      position = retVal.position;
      return parsedElements;
    }


    searchForComments(file){
      commentsLine = this.searchForLineComments(file);
      commentBlock = this.searchForBlockComments(file);
      comments = this.combineComments(commentsLine,commentBlock);
      return comments;
    }
    searchForLineComments(file){
      var commentStack = [];
      file.scanInRange(/(?!^\s+)\/\/.*$/g, file.getRange(), event => commentStack.push(event.range))
      return commentStack;
    }

    searchForBlockComments(file){
      var startBlock = [];
      var endBlock = [];
      var blockStack = [];
      file.scanInRange(/(?!^-+)\/\*.*(!\*|.*$)/g, file.getRange(), event => startBlock.push(event.range));
      file.scanInRange(/(?!^-+)\*\//g, file.getRange(), event => endBlock.push(event.range));
      blockStack = this.createBlockComments(file,startBlock,endBlock);
      return blockStack;
    }

    createBlockComments(file,startBlock,endBlock){
      var loopCtrl = 0;
      blockStack = [];
      while (Object.keys(startBlock).length != 0) {
        if(Object.keys(endBlock).length == 0 && Object.keys(startBlock).length > 0){
          console.log("Adding the end of buffer to end block comment, unended block comment!");
          endBlock.push(file.getRange());
        }
        tempBlock = new Range(startBlock[0].start,endBlock[0].end);
        startBlock.splice(0,1);
        endBlock.splice(0,1);
        startBlock=startBlock.filter(start => !tempBlock.containsRange(start));
        blockStack.push(tempBlock);
        loopCtrl += 1;
        if (loopCtrl == 1000) {
          console.error("Infinite loop for getting block comments.");
          return blockStack;
        }
      }
      return blockStack;
    }

    combineComments(line,block){
      block.forEach(function(blk) {
        line = line.filter(lin => !blk.containsRange(lin));
      });
      comments = line.concat(block);
      comments = comments.sort((a,b) => a.start.row-b.start.row);
      return comments;
    }



}


lib/systemverilog-browser.js
'use babel';

import SystemverilogBrowserView from './systemverilog-browser-view';
import { CompositeDisposable, Disposable } from 'atom';

export default {

  systemverilogBrowserView: null,
  subscriptions: null,

  activate(state) {
    this.systemverilogBrowserView = new SystemverilogBrowserView();
    this.subscriptions = new CompositeDisposable();
    // Add an opener for our view.
    this.subscriptions.add(atom.workspace.addOpener(uri => {
      if (uri === 'atom://systemverilog-browser') {
        view = new SystemverilogBrowserView();
        return view;
      }
    }));


    // Destroy any ActiveEditorInfoViews when the package is deactivated.
    this.subscriptions.add(new Disposable(() => {
      atom.workspace.getPaneItems().forEach(item => {
        if (item instanceof SystemverilogBrowserView) {
          item.destroy();
        }
      });
    }));

    // Register command that toggles this view
    this.subscriptions.add(atom.commands.add('atom-workspace', {'systemverilog-browser:toggle': () => this.toggle()}));
  },

  deactivate() {
    this.subscriptions.dispose();
    this.systemverilogBrowserView.destroy();
  },

  serialize() {
    return {
      systemverilogBrowserViewState: this.systemverilogBrowserView.serialize()
    };
  },

  copy(){
    return this;
  },

  deserializeSystemverilogBrowserView(serialized) {
    return new SystemverilogBrowserView();
  },

  toggle() {
    atom.workspace.toggle('atom://systemverilog-browser');
  }

};


lib/systemverilog-browser-view.js
'use babel';

import SystemVerilogSearch from './system-verilog-search';

export default class SystemverilogBrowserView {

  constructor(serializedState) {
    // Create root element
    this.element = document.createElement('div');
    this.element.classList.add('systemverilog-browser');
    this.element.style.overflow = "auto";
    this.searchObj = new SystemVerilogSearch();

    // Create message element
    const message = document.createElement('div');
    message.classList.add('message');
    this.element.appendChild(message);
    this.subscriptions = atom.workspace.getCenter().observeActivePaneItem(item => {
      if (!atom.workspace.isTextEditor(item)) {
        message.innerText = 'Open a file to see important information about it.';
        return;
      }
      if(!(item.getRootScopeDescriptor() == '.source.systemverilog')){
        message.innerText = 'The file is not systemverilog.';
        message.innerHTML = `
          <h2>The file is not systemverilog</h2>
          <ul>
            <li><b>Works only with systemverilog.</b></li>
          </ul>
        ` ;
        return;
      }
      this.updatePanel(message,item);
      var self = this;
      item.onDidSave(() => { this.refreshFrontEnd(item,self,message); });
    });
  }

    updatePanel(message,editor){
      // this.searchResult = this.searchObj.searchCurrentEditor();
      // this.fileSearch = this.searchObj.searchProject();

      message.innerHTML = ``;

      var self = this;
      const titleElem = document.createElement('h2');
      titleElem.innerHTML = `<h2>SV Elements - ${editor.getTitle()}</h2>`;
      titleElem.style.cursor = "not-allowed";
      titleElem.style.textAlign = "center";
      titleElem.addEventListener("click",function (event) {
        self.destroyVisibleMarkers(editor);
      });
      message.appendChild(titleElem);

      const selector = document.createElement('div');
      // selector.style.textAlign = "center";
      selector.style.paddingBottom = "30px";
      selector.style.paddingLeft = "10px";
      selector.style.textAlign = "center";
      selector.innerHTML = `
        <div class='btn-group'>
          <button class='btn'>Project Browser</button>
          <button class='btn'>File Browser</button>
        </div>
      `;
      self.fileOrProject = 0;
      selector.children[0].lastElementChild.classList.add('selected');
      selector.children[0].firstElementChild.addEventListener("click",function (event) {
        this.classList.add('selected');
        selector.children[0].lastElementChild.classList.remove('selected');
        self.destroyVisibleMarkers(editor);
        self.fileOrProject = 1;
        self.refreshFrontEnd(editor,self,message)
      });
      selector.children[0].lastElementChild.addEventListener("click",function (event) {
        this.classList.add('selected');
        selector.children[0].firstElementChild.classList.remove('selected');
        self.destroyVisibleMarkers(editor);
        self.fileOrProject = 0;
        self.refreshFrontEnd(editor,self,message)
      });
      message.appendChild(selector);
      message.appendChild(document.createElement('div'));
      this.refreshFrontEnd(editor,self,message,this.searchResult,this.fileSearch)
    }

    refreshFrontEnd(editor,self,message){
      // this.fileSearch = this.searchResult;
      message.removeChild(message.lastElementChild);
      const wrapper = document.createElement('div');
      if(this.fileOrProject == 0){
        this.searchResult = this.searchObj.searchCurrentEditor();
        this.displayElements(editor,self,wrapper,this.searchResult,1);
      } else {
        this.fileSearch = this.searchObj.searchProject();
        this.displayElements(editor,self,wrapper,this.fileSearch,1);
      }
      message.appendChild(wrapper)
    }


    destroyVisibleMarkers(item){
      markers = item.getMarkers();
      for (var i = 0; i < Object.keys(markers).length; i++) {
        markers[i].destroy();
      }
    }

    displayElements(editor,self,parentElement,elementList,indent){
      for (var i = 0; i < Object.keys(elementList).length; i++) {
        try {

        (function () {
          var funcId = i;
          const localParent = document.createElement('div');
          const localLine = document.createElement('p');
          const localElem = document.createElement('span');
          const collapseElem = document.createElement('span');
          localElem.innerHTML = `<b>${elementList[funcId].displayString}</b>`;
          localElem.classList.add(elementList[funcId].type.colorClass);
          localElem.style.cursor = "pointer";
          localElem.addEventListener("click", function() { self.funcClick(parentElement,editor,self,elementList[funcId]);}, false);
          localLine.addEventListener("mouseenter",function(){this.classList.add('selected');});
          localLine.addEventListener("mouseleave",function(){this.classList.remove('selected');});
          collapseElem.style.cursor = "pointer";
          if(Object.keys(elementList[funcId].elements).length > 0){
            collapseElem.innerHTML= `<span class="icon icon-chevron-right">`;
            collapseElem.addEventListener("click", function() {
              var content = this.parentElement;
              content = content.nextElementSibling;
              if (content.style.display === "none"){
                this.innerHTML= `<span class="icon icon-chevron-down"></span>`;
                this.style.paddingRight = "0px";
                content.style.display = "block";
              } else {
                this.innerHTML = `<span class="icon icon-chevron-right"></span>`;
                this.style.paddingRight = "0px";
                content.style.display = "none";
              }
            }, false);
          } else {
            collapseElem.innerHTML= `<span></span>`;
            collapseElem.style.paddingRight = "21px";
          }
          localLine.appendChild(collapseElem);
          localLine.appendChild(localElem);
          localParent.appendChild(localLine);
          localParent.style.paddingLeft = (indent*10).toString()+"px";
          if(Object.keys(elementList[funcId].elements).length > 0) {
            const childrenElem = document.createElement('div');
            childrenElem.style.display = "none";
            self.displayElements(editor,self,childrenElem,elementList[funcId].elements,indent+1);
            localParent.appendChild(childrenElem);
          }
          parentElement.appendChild(localParent);
        }());
      } catch (e) {
        console.log(e);
      }
      }
    }

    funcClick(message,editor,self,svElem){
      self.destroyVisibleMarkers(editor);
      if (editor.getPath() == svElem.filePath) {
        editor.scrollToScreenPosition(svElem.startRange.start);
        marker = editor.markBufferRange(svElem.startRange);
        decoration = editor.decorateMarker(marker, {type:'highlight',class:'highlighStyle2'});
      } else {
        atom.workspace.open(svElem.filePath).then(function(newEditor){
          newEditor.scrollToScreenPosition(svElem.startRange.start);
          marker = newEditor.markBufferRange(svElem.startRange);
          decoration = newEditor.decorateMarker(marker, {type:'highlight',class:'highlighStyle2'});
        }).catch(error => console.log(error));
      }
    }

  // Returns an object that can be retrieved when package is activated
  // serialize() {}

  // Returns an object that can be retrieved when package is activated
  serialize() {
    return {
      // This is used to look up the deserializer function. It can be any string, but it needs to be
      // unique across all packages!
      deserializer: 'systemverilog-browser/SystemverilogBrowserView'
    };
  }

  // Tear down any state and detach
  destroy() {
    this.element.remove();
  }

  getElement() {
    return this.element;
  }

  getTitle() {
    // Used by Atom for tab texts
    return 'SystemVerilog Browser';
  }

  getURI() {
    // Used by Atom to identify the view when toggling.
    return 'atom://systemverilog-browser';c
  }

  getDefaultLocation() {
    // This location will be used if the user hasn't overridden it by dragging the item elsewhere.
    // Valid values are "left", "right", "bottom", and "center" (the default).
    return 'left';
  }

  getAllowedLocations() {
    // The locations into which the item can be moved.
    return ['left', 'right', 'bottom'];
  }

}


lib/systemverilog-file-elements.js
'use babel';

import { Range } from 'atom';
import SystemVerilogType from './systemverilog-type';

export default class SystemVerilogCodeElem {

  constructor(startRange,endRange,textString,fileInfo){
    parsedFileInfo = fileInfo.match(/(?!\/)\w*(\.\w*|\w*)$/g)[0].split(".");
    this.fileName = parsedFileInfo[0]+"."+parsedFileInfo[1];
    this.fileExtension = parsedFileInfo[1];
    this.filePath = fileInfo;
    this.set(startRange,endRange,textString);
    this.elements = [];
    this.parent = null;
  }

  setEndRange(endRange){
    this.endRange = endRange;
    this.overallRange = new Range(this.startRange.start,endRange.end);
  }

  set(startRange,endRange,textString){
      this.fingerprint = this.hashCode(startRange.toString() + endRange.toString() + textString);
      this.wholeContent = textString;
      this.startRange = startRange;
      this.endRange = endRange;
      this.overallRange = new Range(startRange.start,endRange.end);
      this.type = new SystemVerilogType(textString);
      this.hardlink = "";
      this.symlink = "";

      filteredStr = textString.match(/(\b.*(?=\#)|\b.*(?=\()|\b.*$)/g);
      this.displayString = filteredStr[0];
      if (this.type.str == "import") {
        temp = textString.match(/\b(\w*)(?=::)/g);
        if(temp == null){
          this.symlink = textString;
        } else {
          this.symlink = temp[0];
        }
      } else if (this.type.str == "include") {
        temp = textString.match(/(?!")\w*\.\w*(?=")/g);
        this.hardlink = temp[0];
      }

  }

  hashCode(s) {
      for(var i = 0, h = 0xdeadbeef; i < s.length; i++)
          h = Math.imul(h ^ s.charCodeAt(i), 2654435761);
      return (h ^ h >>> 16) >>> 0;
  };

  //
  compare(elem){
    stringId = elem.startRange.toString() + elem.endRange.toString() + elem.textString;
    this.testFingerprint = this.hashCode(stringId);
    return this.fingerprint == testFingerprint;
  }


  searchForType(desiredType,stack){
    if(this.type.id == this.type.hashCode(desiredType)){
      stack.push(this);
    }
    if(Object.keys(this.elements).length > 0){
      try {
        this.elements.forEach(function(element){
          element.searchForType(desiredType,stack)
        });
      } catch (e) {
        console.log(e);
        return stack;
      }
    }

    return stack;
  }
  searchForName(desiredName,stack){
    if(this.type.hashCode(this.displayString.split(" ")[1]) == this.type.hashCode(desiredName)){
      stack.push(this);
    }
    if(Object.keys(this.elements).length > 0){
      this.elements.forEach(function(element){
        element.searchForType(desiredName,stack)
      });
    }

    return stack;
  }
  fixParentsInTree(parent){
    this.parent = parent;
    self=this;
    this.elements.forEach(function(element){
      element.fixParentsInTree(self);
    });
  }

}


lib/systemverilog-type.js
'use babel';

export default class SystemVerilogType {

  constructor(s){
    tempStr = s.split(" ");

    this.extern = 0;
    this.virtual = 0;

    if (tempStr[0] == "extern"){
      tempStr.splice(0,1);
      this.extern = 1;
    }

    if (tempStr[0] == "virtual"){
      tempStr.splice(0,1);
      this.virtual = 1;
    }

    testStr = tempStr[0];


    this.withRange = 1;

    greyC = "text-subtle";
    whiteC = "text-highlight";
    blueC = "text-info";
    greenC = "text-success";
    orangeC= "text-warning";
    redC = "text-error";

    if(testStr.substr(0,8)=="include\""){
      testStr = "include";
    }

    switch (testStr) {
      case "import":
        this.withRange = 0;
        this.str = "import";
        this.id = this.hashCode("import");
        this.border = 1;
        this.colorClass = blueC;
        break;
      case "include":
        this.withRange = 0;
        this.str = "include";
        this.id = this.hashCode("include");
        this.border = 1;
        this.colorClass = greyC;
        break;
      case "interface":
        this.str = "interface";
        this.id = this.hashCode("interface");
        this.border = 0;
        this.colorClass = greenC;
        break;
      case "endinterface":
        this.str = "interface";
        this.id = this.hashCode("interface");
        this.border = 1;
        this.colorClass = greenC;
        break;
      case "covergroup":
        this.str = "covergroup";
        this.id = this.hashCode("covergroup");
        this.border = 0;
        this.colorClass = blueC;
        break;
      case "endgroup":
        this.str = "covergroup";
        this.id = this.hashCode("covergroup");
        this.border = 1;
        this.colorClass = blueC;
        break;
      case "module":
        this.str = "module";
        this.id = this.hashCode("module");
        this.border = 0;
        this.colorClass = greenC;
        break;
      case "endmodule":
        this.str = "module";
        this.id = this.hashCode("module");
        this.border = 1;
        this.colorClass = greenC;
        break;
      case "package":
        this.str = "package";
        this.id = this.hashCode("package");
        this.border = 0;
        this.colorClass = whiteC;
        break;
      case "endpackage":
        this.str = "package";
        this.id = this.hashCode("package");
        this.border = 1;
        this.colorClass = whiteC;
        break;
      case "class":
        this.str = "class";
        this.id = this.hashCode("class");
        this.border = 0;
        this.colorClass = blueC;
        break;
      case "endclass":
        this.str = "class";
        this.id = this.hashCode("class");
        this.border = 1;
        this.colorClass = blueC;
        break;
      case "function":
        this.str = "function";
        this.id = this.hashCode("function");
        this.border = 0;
        this.colorClass = blueC;
        break;
      case "endfunction":
        this.str = "function";
        this.id = this.hashCode("function");
        this.border = 1;
        this.colorClass = blueC;
        break;
      case "task":
        this.str = "task";
        this.id = this.hashCode("task");
        this.border = 0;
        this.colorClass = blueC;
        break;
      case "endtask":
        this.str = "task";
        this.id = this.hashCode("task");
        this.border = 1;
        this.colorClass = blueC;
        break;
      case "svFile":
        this.str = "svFile";
        this.id = this.hashCode("svFile");
        this.border = 0;
        this.colorClass = greyC;
        break;
      case "packages":
        this.str = "packages";
        this.id = this.hashCode("packages");
        this.border = 0;
        this.colorClass = greyC;
        break;
      default: this.id = 0; break;

    }
    if(this.extern==1){
      this.border = 1;
    }
  }

  strComp(s){
    if (s == this.str || s == "end"+this.str){
      return true;
    } else {
      return false;
    }
  }

  hashCode(s) {
      for(var i = 0, h = 0xdeadbeef; i < s.length; i++)
          h = Math.imul(h ^ s.charCodeAt(i), 2654435761);
      return (h ^ h >>> 16) >>> 0;
  }

}
